In [1]:
#@title
%%capture
!pip install zipfile36

UsageError: Line magic function `%%capture` not found.


## **Task 1 (20 marks): Why do employee leave?** 

Employees are the backbone of any organization. Its performance is heavily based on the quality of the employees and retaining them. With employee attrition, organizations are faced with a number of challenges ([source](https://github.com/IBM/employee-attrition-aif360)):

* Expensive in terms of both money and time to train new employees
* Loss of experienced employees
* Impact on productivity
* Impact on profit

### **Dataset description** 

The dataset used in this task contains HR analytics data of employees that stay and leave. The types of data include metrics such as education level, job satisfactions, and commmute distance. The goal is to uncover the factors that lead to employee attrition.

### **Data collection**

Below are the codings for some categorial variables in the datasets:

**Education**
1: 'Below College'
2: 'College'
3: 'Bachelor'
4: 'Master'
5: 'Doctor'

**EnvironmentSatisfaction**
1: 'Low'
2: 'Medium'
3: 'High'
4: 'Very High'

**JobInvolvement** 
1: 'Low'
2: 'Medium'
3: 'High'
4: 'Very High'

**JobSatisfaction** 
1: 'Low'
2: 'Medium'
3: 'High'
4: 'Very High'

**PerformanceRating**
1: 'Low'
2: 'Good'
3: 'Excellent'
4: 'Outstanding'

**RelationshipSatisfaction**
1: 'Low'
2: 'Medium'
3: 'High'
4: 'Very High'

**WorkLifeBalance** 
1: 'Bad'
2: 'Good'
3: 'Better'
4: 'Best'

The dataset can be loaded by running the below code cell

In [2]:
import pandas as pd
df_attrition = pd.read_csv("https://raw.githubusercontent.com/IBM/employee-attrition-aif360/master/data/emp_attrition.csv")
df_attrition.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

### **Task description**
**Use visualisations to address the below questions** 

1.1.   Is there a correlation between age and daily rate? (5 marks)

1.2    Is there a difference in attrition between male and femal employees? (5 marks)

1.3   What is the distribution of daily rates across different departments? (5 marks)

1.4  What factors lead to employee attritions? Focusing on four columns: Department, OverTime, JobLevel, and JobSatisfaction (5 marks)

You must write short description below each visualisation to explain "what" you are trying to show and "how" you have chosen to show it.

*Example (not related to this dataset): This visualisation is showing the change in unemployment rate over time. I have used a line chart to show the change over time. Colouring has been used to discriminate between different cities. The city of New York has been highlighted by increasing the thickness of its line.*


**Hint**: attrition rates across different groups can be compared with aggregation in `pandas`. See the example in the code cell below.

In [3]:
# check attrition ratio (yes and no) for each category of EducationField column
df_agg_edu = df_attrition.groupby(["EducationField", "Attrition"]).agg(
    count=pd.NamedAgg(column="Attrition", aggfunc="count")
).reset_index()

df_agg_edu["percent"] = df_agg_edu["count"] / df_agg_edu.groupby(["EducationField"])["count"].transform('sum')

df_agg_edu

EducationField Attrition  count   percent
0    Human Resources        No     20  0.740741
1    Human Resources       Yes      7  0.259259
2      Life Sciences        No    517  0.853135
3      Life Sciences       Yes     89  0.146865
4          Marketing        No    124  0.779874
5          Marketing       Yes     35  0.220126
6            Medical        No    401  0.864224
7            Medical       Yes     63  0.135776
8              Other        No     71  0.865854
9              Other       Yes     11  0.134146
10  Technical Degree        No    100  0.757576
11  Technical Degree       Yes     32  0.242424

In [4]:
import plotly.express as px

1.1 correlation between age and daily rate



In [5]:
fig = px.scatter(df_attrition, x="Age", y="DailyRate", trendline = "ols")
fig

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


1.   This visualisation is showing the correlation between age and daily rate using scatter plot. The changes of daily rate is reflected as age increases. There is also a regression line to demonstrate the overal correlation.
2.   There is no clear correlation between age and daily rate. The regression line appears to be very flat.




1.2 Is there a difference in attrition between male and femal employees?

In [6]:
groupby_attrition_gender = df_attrition.groupby(["Attrition", "Gender"]).agg(
    count=pd.NamedAgg(column="Attrition", aggfunc="count")
).reset_index()


groupby_attrition_gender["percent"] = groupby_attrition_gender["count"] /  groupby_attrition_gender.groupby(["Gender"])["count"].transform('sum') *100

groupby_attrition_gender = groupby_attrition_gender.loc[groupby_attrition_gender['Attrition'] == "Yes"]

groupby_attrition_gender.head()

Attrition  Gender  count    percent
2       Yes  Female     87  14.795918
3       Yes    Male    150  17.006803

In [7]:
fig = px.bar(groupby_attrition_gender, x="Attrition", y = "percent", color="Gender", width=800, height=500, barmode="group")
fig

*   This visualisation is comparing the attrition rate (in percentages) between male and female using barchart. Male is coloured in red and female is coloured in blue.
*   There is only a marginal difference in attrition rate between male and female. The attrition is about 17% for male and 14.8% for female.In summary, the attrition rate for male is slightly higher compared to female.


1.3 What is the distribution of daily rates across different departments?

In [8]:
groupby_dailyrate_dpt = df_attrition[["DailyRate", "Department"]].groupby(["Department"], as_index=False).mean()
groupby_dailyrate_dpt

Department   DailyRate
0         Human Resources  751.539683
1  Research & Development  806.851197
2                   Sales  800.275785

In [9]:
fig = px.bar(groupby_dailyrate_dpt, x="Department", y="DailyRate", width=800, height=500, barmode="group")
fig.update_layout(
    title="Comparing average daily rate between departments",
    xaxis_title="Department",
    yaxis_title="Average daily rate",
)
fig

1.   This visualisation is comparing the attrition rate (in percentages) between departments using barchart.
2.   There is only marginal differences between the 3 departments. Research and development gets paid the most, Sales comes second then Human Resources at last.



In [10]:
fig = px.box(df_attrition, x="Department", y="DailyRate", color="Department")
fig.update_layout(
    title="Comparing daily rate between departments",
    xaxis_title="Department",
    yaxis_title="Daily rate",
)
fig

1.   This visualisation is comparing the distrubition of daily rate (in percentages) between departments using boxplot. Colouring has been used to discriminate between departments. 
2.   According to the box plot, the three departments have very similary minimum, maximum and median for daily rate.



In [11]:
fig = px.histogram(df_attrition, x="DailyRate", facet_row="Department", color="Department",
                   opacity=0.5, nbins=28, histnorm='percent',
                   title="Distribution of daily rate between departments")
fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[0]))
fig

1.   This visualisation is comparing the distrubition of daily rate (in percentages) between departments using histogram. Colouring has been used to discriminate between departments. The graph is faceted by department.
2.   According to the histogram, the 3 departments have very similar distrubitions, there are only a few negligible differences. 1 thing to notice is that the number of HR employees is a lot smaller than the other 2 departments, therefore the variability of daily rate is bigger.


1.4 What factors lead to employee attritions? Focusing on four columns: Department, OverTime, JobLevel, and JobSatisfaction

In [12]:
groupby_attr_dpt = df_attrition.groupby(["Attrition", "Department"]).agg(
    count=pd.NamedAgg(column="Attrition", aggfunc="count")
).reset_index()

groupby_attr_dpt["percent"] = groupby_attr_dpt["count"] /  groupby_attr_dpt.groupby(["Department"])["count"].transform('sum') *100

groupby_attr_dpt = groupby_attr_dpt.loc[groupby_attr_dpt['Attrition'] == "Yes"]

groupby_attr_dpt

Attrition              Department  count    percent
3       Yes         Human Resources     12  19.047619
4       Yes  Research & Development    133  13.839750
5       Yes                   Sales     92  20.627803

In [13]:
fig = px.bar(groupby_attr_dpt, x="Attrition", y="percent", color="Department", barmode = "group")
fig

1.   This visualisation is comparing the attrition rate (in percentages) between departments using barchart. Colouring has been used to discriminate between departments.
2.   The rate of attrition accross three departments are generally below 21%. However, research and developement has the lowest attrition rate at about 14% as supposed to sales with highest attrition rate at 20.6% and HR at about 19%. 



In [14]:
groupby_attr_ot = df_attrition.groupby(["Attrition", "OverTime"]).agg(
    count=pd.NamedAgg(column="Attrition", aggfunc="count")
).reset_index()

groupby_attr_ot["percent"] = groupby_attr_ot["count"] /  groupby_attr_ot.groupby(["OverTime"])["count"].transform('sum') *100

groupby_attr_ot = groupby_attr_ot.loc[groupby_attr_ot['Attrition'] == "Yes"]

groupby_attr_ot

Attrition OverTime  count    percent
2       Yes       No    110  10.436433
3       Yes      Yes    127  30.528846

In [ ]:
fig = px.bar(groupby_attr_ot, x="Attrition", y="percent", color="OverTime", barmode = "group")
fig

1.   This visualisation is comparing the attrition rate (in percentages) between overtime categories using barchart. Colouring has been used to discriminate between overtime categories.
2.   It is clear that attrition rate is about 3 times higher for those who have to work overtime.



In [15]:
df_attrition['JobLevel']=df_attrition.JobLevel.astype('category')

In [16]:
groupby_attr_jl = df_attrition.groupby(["Attrition", "JobLevel"]).agg(
    count=pd.NamedAgg(column="Attrition", aggfunc="count")
).reset_index()

groupby_attr_jl["percent"] = groupby_attr_jl["count"] /  groupby_attr_jl.groupby(["JobLevel"])["count"].transform('sum') *100

groupby_attr_jl = groupby_attr_jl.loc[groupby_attr_jl['Attrition'] == "Yes"]

groupby_attr_jl

Attrition JobLevel  count    percent
5       Yes        1    143  26.335175
6       Yes        2     52   9.737828
7       Yes        3     32  14.678899
8       Yes        4      5   4.716981
9       Yes        5      5   7.246377

In [17]:
fig = px.bar(groupby_attr_jl, x="Attrition", y="percent", color="JobLevel", barmode = "group")
fig

1.   This visualisation is comparing the attrition rate (in percentages) between job level categories using barchart. Colouring has been used to discriminate between job level categories.
2.   Generally speaking, the lower the job level, the higher the attrition rate. Job level 1 has the highest attrition rate at about 26%, followed by job level 3, level 2 and level 5 in descending order. Job level 4 has the lowest attrition rate.



In [18]:
df_attrition['JobSatisfaction']=df_attrition.JobSatisfaction.astype('category')

In [19]:
groupby_attr_js = df_attrition.groupby(["Attrition", "JobSatisfaction"]).agg(
    count=pd.NamedAgg(column="Attrition", aggfunc="count")
).reset_index()

groupby_attr_js["percent"] = groupby_attr_js["count"] /  groupby_attr_js.groupby(["JobSatisfaction"])["count"].transform('sum') *100

groupby_attr_js = groupby_attr_js.loc[groupby_attr_js['Attrition'] == "Yes"]


groupby_attr_js

Attrition JobSatisfaction  count    percent
4       Yes               1     66  22.837370
5       Yes               2     46  16.428571
6       Yes               3     73  16.515837
7       Yes               4     52  11.328976

In [20]:
fig = px.bar(groupby_attr_js, x="Attrition", y="percent", color="JobSatisfaction", barmode = "group")
fig

1.   This visualisation is comparing the attrition rate (in percentages) between job satisfaction categories using barchart. Colouring has been used to discriminate between job satisfaction categories.
2.   Generally speaking, the higher the job satisfaction, the lower attrition rate is. Those with low job satisfaction has the highest attrition rate at about 23%. Those with medium and high job satisfaction rate are at a tie at about 16.5% attrition rate. Those with very high job satisfaction has the lowest attrition rate at only 11.3%.

